docs/_posts/2017-12-31-ldfi.md

```diff
@@ -0,0 +1,210 @@
+
+Molly list the faults that any general-purpose verifyer should have. 
+I will focus on the RPC general-purpose verifyer as an abstraction layer 
+such that with zero work you get a state of the art fault-injector. 
+
+* Message loss: `drop n out of N`
+* Crash failures: `sys.exit(1)`
+* Nondeterminisim in ordering and timing: `add arbitrary delayes per message(rpc) same as above for RPC` 
+
+Note that this system is for correctness and not performance. 
+
+# `smf::condor` codegenerated LDFI for a given fspec
+
+* Main assumption: 
+
+  * Succesfully delivered messages are received in a deterministic order
+  * systematically explore failures.
+
+
+Failures explored:
+
+* permanent crashed failrues 
+* message loss 
+* network partitions. 
+
+Failures not explored:
+
+* byzantine failures
+* crash-recovery
+
+## Data model
+
+```
+struct fspec {
+  // logical bound of executions. A counter.
+  // end_of_time
+  eot: uint = 0;
+  
+  // used for failure quiescence. That is, let the
+  // system recover 
+  // logical time at which message loss stops.
+  //
+  // eot < eff - always
+  //
+  // end_of_finite_failures
+  eff: uint = 0;
+  
+  // number of times to call sys.exit(1)
+  // Process crash counter.
+  //
+  crashes: uint = 0;
+}
+
+struct ldfi_test {
+   // a duration of 0 means do not explore,
+   // instead use the manually loaded spec
+   // to exercise the system
+   //
+   duration_ms: uint = 0;
+   spec:        fspec;
+}
+
+
+```
+
+
+For trivial protocols like request-response, etc., 
+we can model the protocol at the code generation level. 
+
+For more complex protocols with harbeats, keepalives, 
+anti-entropy, timeouts, retries, etc, the system 
+programmer must extend some interface like such:
+
+```
+
+// basically any callable. - need to figure out input
+// and output type
+//
+//
+template<typename Input, 
+         typename... Outputs>
+struct ldfi_backward_step {
+    seastar::future<...Outputs> operator()(Input i) {}
+}
+
+```
+
+
+Currently, the proposal does not cover the _hazard step_
+of Molly. This is deferred to the second phase of implementation.
+
+
+## Implementation proposal
+
+`smf::rpc` comes with a code generation program `smf_gen`. Currently
+it generates RPC stubs like Thrift, gRPC, Cap 'n Proto, etc. The idea
+is to extend this code generation step behind a flag of `--ldfi`.
+
+Because we understand the inputs and outputs (flatbuffers c-structs),
+as well as the actual communication infrastructure (`smf::rpc`), we can 
+generate a proxy that can do the basic accounting. As such, ALL communications
+for a system being tested will go through a single process proxy.
+
+This proxy `smf::condor` will control:
+
+* Control message loss: by simply dropping messages (not proxying)
+* Delivery ordering and timing: by adding arbitrary `future<> sleep()` delays
+to messages
+
+In addition of this proxy, smf needs to provide filters to do the actual crashing
+of processes via `exit(1)`.
+
+
+```
+
+namespace smf {
```

Comment:
This has a dependency on finally adding header parsing which we've ignored after the fbs issue of nested fbs'  - don't think we've chained the parsing yet.